### PR TITLE
Simplify by removing formats and registration

### PR DESCRIPTION
**Pull request by tfpauly**
- Remove datagram format registry -- new capsule types can be used instead
- Delegate context ID registration to methods/protocols, which can define new capsule types
- Remove the need for header negotiation to support contexts. Individual methods can define their own headers for negotiation.
- Datagram contexts don't always need to be used, but there's an implicit "context 0" which has a format that must be defined by each method. The presence of a context ID in a datagram is encoded in the "encoded stream ID".

[Rendered view of this PR](https://ietf-wg-masque.github.io/draft-ietf-masque-h3-datagram/tfp/simplify/draft-ietf-masque-h3-datagram.html).

### CODE REVIEW

File: draft-ietf-masque-h3-datagram.md [hunk 1]
```diff
@@ -180,56 +161,57 @@ of {{QUIC}}):
...
 Context ID:
 
 : A variable-length integer indicating the context ID of the datagram (see
 {{datagram-contexts}}). Whether or not this field is present depends on whether
-datagram contexts are in use on this stream, see {{context-hdr}}. If this QUIC
-DATAGRAM frame is reordered and arrives before the receiver knows whether
-datagram contexts are in use on this stream, then the receiver cannot parse
-this datagram and the receiver MUST either drop that datagram silently or
-buffer it temporarily.
+the least-significant bit of the Encoded Stream ID, as described above. If the
```
Comment (DavidSchinazi) [inline draft-ietf-masque-h3-datagram.md:190]:
```suggestion
the least-significant bit of the Encoded Stream ID is set, as described above. If the
```

File: draft-ietf-masque-h3-datagram.md [hunk 2]
```diff
@@ -180,56 +161,57 @@ of {{QUIC}}):
...
+receiver of a datagram does not support Context IDs, it will drop any datagram
+that contains a context ID.
```
Comment (DavidSchinazi) [inline draft-ietf-masque-h3-datagram.md:192]: Should we say something like: `If the Context ID field is not present, this datagram is associated with the default context (which has its Context ID set to 0).`

File: draft-ietf-masque-h3-datagram.md [hunk 3]
```diff
@@ -919,47 +627,33 @@ STREAM(44): HEADERS            -------->
...
                         :status = 200
-                        Sec-Use-Datagram-Contexts = ?1
+                        Sec-UDP-Timestamps = ?1
 
 STREAM(44): DATA               -------->
-  Capsule Type = REGISTER_DATAGRAM_CONTEXT
-  Context ID = 0
-  Datagram Format Type = UDP_PAYLOAD
-  Datagram Format Additional Data = ""
+  Capsule Type = REGISTER_UDP_WITH_TIMESTAMP_CONTEXT
```
Comment (DavidSchinazi) [inline draft-ietf-masque-h3-datagram.md:637]: In this example, the client wants to send the datagram without the extension to make sure that its first packet gets through to the target. Same for the next example below

File: draft-ietf-masque-h3-datagram.md [hunk 4]
```diff
@@ -74,10 +70,8 @@ This document is structured as follows:
 * {{capsule}} introduces the Capsule Protocol and the "data stream" concept.
   Data streams are initiated using special-purpose HTTP requests, after which
   Capsules, an end-to-end message, can be sent.
-  * The following Capsule types are defined, together with guidance for defining new types:
-    * Datagram registration capsules {{register-capsule}}
-    * Datagram close capsule {{close-capsule}}
-    * Datagram capsules {{datagram-capsule}}
+  * Datagram Capsule types {{datagram-capsule}} are defined, along with guidance
+    for defining new capsule types.
```
Comment (LPardue) [inline draft-ietf-masque-h3-datagram.md:74]:
```suggestion
  * {{datagram-capsule}} defines Datagram Capsule types, along with guidance
    for specifying new capsule types.
```

File: draft-ietf-masque-h3-datagram.md (hunk 2 above)
Comment (tfpauly) [reply draft-ietf-masque-h3-datagram.md:192]:
```suggestion
that contains a context ID. If the Context ID field is not present, this datagram
is associated with the default context (which has its Context ID set to 0).
```

File: draft-ietf-masque-h3-datagram.md (hunk 3 above)
Comment (tfpauly) [reply draft-ietf-masque-h3-datagram.md:645]: @DavidSchinazi I've updated the examples—now it just has a "supported extension" and "unsupported extension" case.

File: draft-ietf-masque-h3-datagram.md [hunk 5]
```diff
@@ -583,6 +404,32 @@ loss; this can misrepresent the true path properties, defeating methods such a
 DPLPMTUD.
 
 
+## Registering Datagram Contexts with Capsules {#register-capsules}
+
+Methods or protocols that support multiple datagram payload formats or
```
Comment (LPardue) [inline draft-ietf-masque-h3-datagram.md:409]: maybe qualify "methods or protocols" here?

File: draft-ietf-masque-h3-datagram.md [hunk 6]
```diff
@@ -919,98 +629,38 @@ STREAM(44): HEADERS            -------->
...
-           <--------  STREAM(44): HEADERS
-                        :status = 200
-                        Sec-Use-Datagram-Contexts = ?1
-
-STREAM(44): DATA               -------->
-  Capsule Type = REGISTER_DATAGRAM_CONTEXT
-  Context ID = 0
-  Datagram Format Type = UDP_PAYLOAD
-  Datagram Format Additional Data = ""
+  Sec-UDP-Timestamps = ?1
```
Comment (LPardue) [inline draft-ietf-masque-h3-datagram.md:632]: nit: lowercase

```suggestion
  sec-udp-timestamps = ?1
```

File: draft-ietf-masque-h3-datagram.md [hunk 7]
```diff
@@ -180,56 +161,58 @@ of {{QUIC}}):
 
 ~~~
 HTTP/3 Datagram {
-  Quarter Stream ID (i),
+  Encoded Stream ID (i),
```
Comment (DavidSchinazi) [inline draft-ietf-masque-h3-datagram.md:164]: Since we're changing the wire format, we need to bump the value of the SETTING

File: draft-ietf-masque-h3-datagram.md [hunk 8]
```diff
@@ -773,11 +556,8 @@ This registry initially contains the following entries:
 
 | Capsule Type                 |   Value   | Specification |
 |:-----------------------------|:----------|:--------------|
-| REGISTER_DATAGRAM_CONTEXT    | 0xff37a1  | This Document |
-| REGISTER_DATAGRAM            | 0xff37a2  | This Document |
-| CLOSE_DATAGRAM_CONTEXT       | 0xff37a3  | This Document |
-| DATAGRAM_WITH_CONTEXT        | 0xff37a4  | This Document |
-| DATAGRAM                     | 0xff37a5  | This Document |
+| DATAGRAM                     | 0xff37a1  | This Document |
```
Comment (DavidSchinazi) [inline draft-ietf-masque-h3-datagram.md:559]: Please don't change the value, that breaks backwards compatibility

File: draft-ietf-masque-h3-datagram.md [hunk 9]
```diff
@@ -902,13 +607,18 @@ DATAGRAM                        -------->
 
 In these examples, the client supports a CONNECT-UDP Timestamp Extension, which
 uses a different Datagram Format Type that carries a timestamp followed by the
-encapsulated UDP payload.
+encapsulated UDP payload. Datagrams on the default context (0) are expected to
+contain UDP payloads, while contexts established with the
+REGISTER_UDP_WITH_TIMESTAMP_CONTEXT capsule include a timestamp along with the
```
Comment (DavidSchinazi) [inline draft-ietf-masque-h3-datagram.md:612]:
```suggestion
REGISTER_UDP_WITH_TIMESTAMP_CONTEXT capsule include a timestamp followed by the
```

File: draft-ietf-masque-h3-datagram.md (hunk 3 above)
Comment (LPardue) [reply draft-ietf-masque-h3-datagram.md:645]: but if this is the support casem then wouldn't the server reply with `Encoded Stream ID = 23` and `Payload = Encapsulated UDP Payload With Timestamp`?

Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:645]: My comment still stands: the client is trying to send without the extension in its first flight

Comment (tfpauly) [reply draft-ietf-masque-h3-datagram.md:645]: It can! It was just getting long. I'll fix.

File: draft-ietf-masque-h3-datagram.md (hunk 7 above)
Comment (tfpauly) [reply draft-ietf-masque-h3-datagram.md:164]: Done!

File: draft-ietf-masque-h3-datagram.md (hunk 8 above)
Comment (tfpauly) [reply draft-ietf-masque-h3-datagram.md]: Whoops, fixed

File: draft-ietf-masque-h3-datagram.md (hunk 6 above)
Comment (tfpauly) [reply draft-ietf-masque-h3-datagram.md]: Fixed

File: draft-ietf-masque-h3-datagram.md (hunk 5 above)
Comment (tfpauly) [reply draft-ietf-masque-h3-datagram.md]: Qualified in the long form explanation

File: draft-ietf-masque-h3-datagram.md (hunk 3 above)
Comment (tfpauly) [reply draft-ietf-masque-h3-datagram.md:645]: Fixed

File: draft-ietf-masque-h3-datagram.md [hunk 10]
```diff
@@ -180,56 +161,58 @@ of {{QUIC}}):
...
+
+: A variable-length integer that contains an encoded value indicating the
+client-initiated bidirectional stream that this datagram is associated with.
+The Encoded Stream ID for a datagram without a context ID is the associated
+stream ID divided by two. The Encoded Stream ID for a datagram with a context
+ID is the associated stream ID divided by two, plus one. (Client-initiated
+bidirectional streams have stream IDs that are divisible by four, so the
+stream ID divided by two always has the least-significant bit set to zero).
+The largest legal QUIC stream ID value is 2<sup>62</sup>-1, so the largest legal
+value of Encoded Stream ID is (2<sup>62</sup>-1 / 2) + 1. Receipt of a frame that
```
Comment (martinthomson) [inline draft-ietf-masque-h3-datagram.md:181]: Check operator precedence rules.  Also, I simplified this for you.

```suggestion
value of Encoded Stream ID is 2<sup>61</sup>. Receipt of a frame that
```

File: draft-ietf-masque-h3-datagram.md [hunk 11]
```diff
@@ -180,56 +161,58 @@ of {{QUIC}}):
...
-Receipt of a frame that includes a larger value MUST be treated as a connection
-error of type FRAME_ENCODING_ERROR.
+Encoded Stream ID:
+
+: A variable-length integer that contains an encoded value indicating the
+client-initiated bidirectional stream that this datagram is associated with.
+The Encoded Stream ID for a datagram without a context ID is the associated
+stream ID divided by two. The Encoded Stream ID for a datagram with a context
+ID is the associated stream ID divided by two, plus one. (Client-initiated
+bidirectional streams have stream IDs that are divisible by four, so the
```
Comment (martinthomson) [inline draft-ietf-masque-h3-datagram.md:178]: so what does a value of 0 mean?  why divisible by four, but only divide by two?  how is this related to the other simplifications?

**Review by martinthomson — COMMENTED** [review]
Probably OBE, but I don't understand the change from /4 to /2+1.

**Comment by DavidSchinazi** [comment]
Closing in favor of #135
